### PR TITLE
Re-write DOS_Shell::CMD_RENAME

### LINKED ISSUE #2212: Wolfenstein 3D Super Upgrades crash
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

_No response_

### What Operating System are you using?

Windows 11

### If Other OS, please describe

_No response_

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

I installed Wolfenstein 3D v1.4 and it works. I found out about this collection of levels called `Wolfenstein 3D Super Upgrades`. The installation is successful, I run the WolfMaster program (wolfm.bat), choose levels, start the game and it crashes/freezes without errors. This is what I found on a forum:
```
The addon was released as freeware by 3DRealms, and is available (with no support) on their FTP server. Wolf3D is crashing because the external menu program tries to rename some files using wildcards, which DOSBox does not support.

You can work around the situation by using the 4DOS command shell, which has more full-featured versions of DOS commands. Place 4DOS.COM in the root of your mounted C: drive, and then type "SET COMSPEC=C:\4DOS.COM" in DOSBox prior to running the addon. You can put the SET command in the autoexec section of your conf file for convenience. You should also probably re-install Wolf3D and the addon because of the leftover files that didn't get renamed.

```

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Install Wolfenstein 3D v1.4
2. Install Wolfenstein 3D Super Upgrades
3. In the Wolfenstein 3D folder run "setup" (it will install WolfMaster)
4. Run WolfMaster (wolfm.bat), choose some levels and start the game

### Your configuration

_No response_

### Provide a Log

_No response_

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description
Re-wrote the Rename command.  It now supports wildcards and uses C++ style string manipulation.

## Related issues
Fixes #2212

# Manual testing
Confirmed the bug in #2212 is fixed.  Also manually compared some edge cases to real MS-DOS 6.22 (running in 86box) and it mostly behaves the same.  This could probably use some unit tests.  I'm not sure if we have the ability to test shell commands in a "fake" filesystem.  I can look at that tomorrow or something (getting a bit tired of programming for the day :sleeping: and wanted to go ahead and get this up for review.)

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/shell/shell_cmds.cpp [hunk 1]
```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
 	return files;
 }
 
+struct FileExt {
+	std::string root      = {};
+	std::string extension = {};
+};
+
+static FileExt split_extention(const std::string& fullname)
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2500]: typo: extension

File: src/shell/shell_cmds.cpp [hunk 2]
```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
...
+	FileExt ret = {};
+	size_t pos  = fullname.rfind('.');
+	ret.root    = fullname.substr(0, pos);
+	++pos;
+	if (pos > 0 && pos < fullname.size()) {
+		ret.extension = fullname.substr(pos);
+	} else {
+		ret.extension = {};
+	}
+	return ret;
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2511]: Can't we use std::filesystem functions for this? Or we want to make sure we make something very DOS specific here?

File: src/shell/shell_cmds.cpp [hunk 3]
```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
...
+}
+
+static std::string handle_wildcards(const std::string& wild, const std::string& old)
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2514]: wild -> wildcards
old -> old what?

File: src/shell/shell_cmds.cpp [hunk 4]
```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
...
+	}
+
+	// Second argument must not contain a path
+	if (second_arg.find_first_of("\\:") != std::string::npos) {
+		SyntaxError();
+		return;
+	}
+
+	const std::string path = first_arg.substr(0, first_arg.rfind('\\') + 1);
+	const FileExt wild     = split_extention(second_arg);
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2559]: wildcards?

File: src/shell/shell_cmds.cpp [hunk 5]
```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
 	return files;
 }
 
+struct FileExt {
+	std::string root      = {};
+	std::string extension = {};
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2497]: Suggest going with `name` and `ext` instead. "root" has other connotations, and "name" + "extension" is the standard MS-DOS nomenclature.

Comment (johnnovak) [reply src/shell/shell_cmds.cpp:2497]: Also, maybe call this `DosFilename` or something and move it into some common DOS command / FAT / filesystem related header? (I don't deal with DOS filesystem stuff often, so you'll need to research this a bit 😄 )

**Comment by johnnovak** [comment]
> Also manually compared some edge cases to real MS-DOS 6.22 (running in 86box) and it mostly behaves the same.

Could you paste some of these test commands into the manual testing section for our benefit and posterity?

> This could probably use some unit tests. I'm not sure if we have the ability to test shell commands in a "fake" filesystem. I can look at that tomorrow or something (getting a bit tired of programming for the day 😴 and wanted to go ahead and get this up for review.)

@MeAreJeenius has started a very helpful DOS command regression testing page on the wiki, I'd suggest to extend that:
https://github.com/dosbox-staging/dosbox-staging/wiki/Manual-DOS-CLI-Testing

Mocking out a "fake" filesystem would be a bit too rich, I think 😄 We're not OS devs, and it would be a massive effort to maintain such a mocking framework, and the level of confidence of the test results would still not be super high IMO 😅

File: src/dos/dos_files.cpp [hunk 6]
```diff
@@ -815,7 +815,23 @@ bool DOS_Canonicalize(const char* const name, char* const big)
...
-bool DOS_GetFreeDiskSpace(uint8_t drive,uint16_t * bytes,uint8_t * sectors,uint16_t * clusters,uint16_t * free) {
+std::string DOS_Canonicalize(const char* const name)
+{
+	uint8_t drive;
+	char fullname[DOS_PATHLENGTH];
+	if (!DOS_MakeName(name, fullname, &drive)) {
+		return {};
+	}
+	std::string ret;
+	ret.push_back(drive + 'A');
```
Comment (kcgen) [inline src/dos/dos_files.cpp:826]: There's a support.h function:

`char drive_letter(uint8_t index)`

It asserts that the drive index is legal before giving back the drive letter

File: src/shell/shell_cmds.cpp [hunk 7]
```diff
@@ -2532,6 +2492,98 @@ static std::vector<std::string> search_files(const std::string_view query)
...
+		std::string old_fullpath = path + old_filename;
+
+		std::string new_fullpath = path + new_split.root;
+		if (!new_split.extension.empty()) {
+			new_fullpath.push_back('.');
+			new_fullpath.append(new_split.extension);
+		}
+
+		if (!DOS_Rename(old_fullpath.c_str(), new_fullpath.c_str())) {
+			WriteOut("Rename %s -> %s failed\n",
```
Comment (kcgen) [inline src/shell/shell_cmds.cpp:2580]: Question (not asking for changes, not sure of the answer) --
In real DOS, I wonder if a failed rename sets `ERRORLEVEL` that a subsequent batch statement can check? (or something like that?) - maybe that happens at another level though (the filesystem calls?).

**Review by kcgen — COMMENTED** [review]
Just a couple minor comments.
Looking forward to this big addition!

**Comment by FeralChild64** [comment]
@weirddan455 Did you also check behavior on hidden/system/read-only files and directories?

**Comment by Torinde** [comment]
Hidden files are ignored, output file keeps the ARS attributes of the input file, there are no switches about file attributes - per my testing at #2956

I haven't tested if renaming a file after a date change has the date bug #620

**Comment by weirddan455** [comment]
@FeralChild64  I will check that behavior tomorrow but I expect to see no difference because it's using the exact same function call `DOS_Rename` to do the actual renaming as the old code was.  Improvements are in parsing and wildcard handling.

File: src/shell/shell_cmds.cpp (hunk 1 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp]: fixed

File: src/shell/shell_cmds.cpp (hunk 2 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp:2506]: I think it's best to avoid `std::filesystem` when interacting with the DOS side of things.  As I understand it, those functions are meant for abstracting host side differences (ex. on Unix-like systems it may replace DOS-style path separators `\` with Unix-style `/` which we most certainly do not want).

File: src/shell/shell_cmds.cpp (hunk 3 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp]: renamed

File: src/shell/shell_cmds.cpp (hunk 4 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp]: renamed

File: src/shell/shell_cmds.cpp (hunk 5 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp]: Renamed and moved to `dos_system.h`.  Could not find a pre-existing struct for this.  Closest I saw was `DOS_File` but that's a heavy weight class with a bunch of virtual functions and such.

File: src/dos/dos_files.cpp [hunk 8]
```diff
@@ -815,7 +815,23 @@ bool DOS_Canonicalize(const char* const name, char* const big)
 	return true;
 }
 
-bool DOS_GetFreeDiskSpace(uint8_t drive,uint16_t * bytes,uint8_t * sectors,uint16_t * clusters,uint16_t * free) {
+std::string DOS_Canonicalize(const char* const name)
+{
+	uint8_t drive;
```
[the remark this answers is not in the document]
Comment (weirddan455) [reply src/dos/dos_files.cpp]: I copy + pasted this code from a pre-existing `DOS_Caonicalize` function.  This is an overload that does the same thing but returns a `std::string`.  I just pushed a separate commit to clean this up though.  Let me know if that looks better.

File: src/dos/dos_files.cpp (hunk 6 above)
Comment (weirddan455) [reply src/dos/dos_files.cpp]: Using this in latest commit

File: src/shell/shell_cmds.cpp (hunk 7 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp:2575]: `DOS_SetError` gets called inside `DOS_Rename` if it fails.

**Comment by weirddan455** [comment]
Addressed all the code commits.  I'll do more testing later.  @kcgen looks like your Mac doing the CI builds ran out of space.

**Review by kcgen — APPROVED** [review]

File: src/shell/shell_cmds.cpp (hunk 2 above)
Comment (johnnovak) [reply src/shell/shell_cmds.cpp:2506]: I think you-re right, it would be asking for trouble.

File: include/dos_system.h [hunk 9]
```diff
@@ -93,6 +93,11 @@ struct FileStat_Block {
 
 class DOS_DTA;
 
+struct DOS_Filename {
```
Comment (johnnovak) [inline include/dos_system.h:96]: The MODULE_Function naming scheme is really only for API calls. For structs, we drop the module prefix and go with a sensible name.

In this case, the sensible name would still be `DosFilename`, just it would be all camel case like the rest of the structs.

File: src/dos/dos_files.cpp [hunk 10]
```diff
@@ -804,21 +804,40 @@ bool DOS_SetFileAttr(const char* const name, FatAttributeFlags attr)
...
+	// doubt it :)
+
+	// Initalize to invalid drive index so assert will fire if DOS_MakeName
+	// does not set this
+	uint8_t drive                 = 255;
+	char fullname[DOS_PATHLENGTH] = {};
+	if (!DOS_MakeName(name, fullname, &drive)) {
+		return false;
+	}
+	big[0] = drive_letter(drive);
```
Comment (johnnovak) [inline src/dos/dos_files.cpp:816]: Could you rename `big` to something meaningful?

File: src/dos/dos_files.cpp [hunk 11]
```diff
@@ -804,21 +804,40 @@ bool DOS_SetFileAttr(const char* const name, FatAttributeFlags attr)
...
+std::string DOS_Canonicalize(const char* const name)
+{
+	char fullname[DOS_PATHLENGTH] = {};
+	if (!DOS_Canonicalize(name, fullname)) {
+		return {};
+	}
+	return std::string(fullname);
+}
+
+bool DOS_GetFreeDiskSpace(uint8_t drive, uint16_t* bytes, uint8_t* sectors,
```
Comment (johnnovak) [inline src/dos/dos_files.cpp:832]: Const args?

File: src/dos/dos_files.cpp [hunk 12]
```diff
@@ -804,21 +804,40 @@ bool DOS_SetFileAttr(const char* const name, FatAttributeFlags attr)
...
+                          uint16_t* clusters, uint16_t* free)
+{
+	if (drive == 0) {
+		drive = DOS_GetDefaultDrive();
+	} else {
+		drive--;
+	}
```
Comment (johnnovak) [inline src/dos/dos_files.cpp:839]: `--drive` by convention

File: src/shell/shell_cmds.cpp [hunk 13]
```diff
@@ -2532,6 +2492,93 @@ static std::vector<std::string> search_files(const std::string_view query)
...
+	} else {
+		ret.ext = {};
+	}
+	return ret;
+}
+
+static std::string handle_wildcards(const std::string& wildcards,
+                                    const std::string& old_filename)
+{
+	std::string ret = {};
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2512]: Maybe call it what it is? `expanded_name`?

File: src/shell/shell_cmds.cpp [hunk 14]
```diff
@@ -2532,6 +2492,93 @@ static std::vector<std::string> search_files(const std::string_view query)
...
+	}
+	return ret;
+}
+
+void DOS_Shell::CMD_RENAME(char* args)
+{
+	HELP("RENAME");
+
+	const std::string first_arg  = DOS_Canonicalize(strip_word(args));
+	const std::string second_arg = strip_word(args);
```
Comment (johnnovak) [inline src/shell/shell_cmds.cpp:2542]: Can't you use more descriptive names than first and second? Like source and target?

**Comment by johnnovak** [comment]
Couple of more minor remarks @weirddan455 and can you please add manual testing info to the wiki as mentioned? Cheers. Otherwise we're 90% there, great job 🎉

File: src/dos/dos_files.cpp (hunk 11 above)
Comment (weirddan455) [reply src/dos/dos_files.cpp:832]: I didn't make changes to this function.  The format script fixed formatting on it.  I don't think these can be const either since they're pointers that get modified.

File: include/dos_system.h (hunk 9 above)
Comment (weirddan455) [reply include/dos_system.h]: Renamed

File: src/dos/dos_files.cpp (hunk 10 above)
Comment (weirddan455) [reply src/dos/dos_files.cpp]: Renamed

File: src/dos/dos_files.cpp (hunk 12 above)
Comment (weirddan455) [reply src/dos/dos_files.cpp:839]: Again, I didn't write this function but I went ahead and changed that.

File: src/shell/shell_cmds.cpp (hunk 13 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp:2512]: Renamed

Comment (weirddan455) [reply src/shell/shell_cmds.cpp:2512]: Or wait... I renamed the wrong one... 1 sec lol

File: src/shell/shell_cmds.cpp (hunk 14 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp]: Renamed

File: src/shell/shell_cmds.cpp (hunk 13 above)
Comment (weirddan455) [reply src/shell/shell_cmds.cpp]: OK, now it's renamed :sweat_smile:

**Comment by weirddan455** [comment]
>  Mocking out a "fake" filesystem would be a bit too rich, I think 😄 We're not OS devs, and it would be a massive effort to maintain such a mocking framework, and the level of confidence of the test results would still not be super high IMO 😅

Yeah, I just couldn't remember what we had set up in our unit testing framework.  I wrote a unit test for the `MOVE` command a while back but it only tests a single parsing function I wrote, not the entire command.  This command didn't need a full parser like MOVE did so it's a bit trickier to write a unit test.

I added the manual testing to the Wiki though.  Confirmed behavior of those commands matches MS-DOS 6.22.  https://github.com/dosbox-staging/dosbox-staging/wiki/Manual-DOS-CLI-Testing

File: src/dos/dos_files.cpp (hunk 12 above)
Comment (johnnovak) [reply src/dos/dos_files.cpp:839]: Yeah, it's optional, but it's nice to clean and a few minor things when you touch something.

**Review by johnnovak — APPROVED** [review]
Very nice job @weirddan455 , also 🥇 for adding the manual testing steps to the wiki 🎉
